spec.md

```diff
@@ -1021,6 +1056,150 @@ It is NOT REQUIRED for a controller plugin to implement the `LIST_VOLUMES` capab
+If the corresponding Controller Plugin has `PUBLISH_UNPUBLISH_VOLUME` controller capability and the Node Plugin has `STAGE_UNSTAGE_VOLUME` capability, then the CO MUST guarantee that this RPC is called after `ControllerPublishVolume` is called for the given volume on the given node and returns a success.
+The CO MUST guarantee that this RPC is called and returns a success before any `NodePublishVolume` is called for the given volume on the given node.
+This operation MUST be idempotent.
+If this RPC failed, or the CO does not know if it failed or not, it MAY choose to call `NodeStageVolume` again, or choose to call `NodeUnstageVolume`.
+
+```protobuf
+message NodeStageVolumeRequest {
+  // The API version assumed by the CO. This is a REQUIRED field.
+  Version version = 1;
+
+  // The ID of the volume to publish. This field is REQUIRED.
+  string volume_id = 2;
+
+  // The CO SHALL set this field to the value returned by
+  // `ControllerPublishVolume` if the corresponding Controller Plugin
+  // has `PUBLISH_UNPUBLISH_VOLUME` controller capability, and SHALL be
+  // left unset if the corresponding Controller Plugin does not have
+  // this capability. This is an OPTIONAL field.
+  map<string, string> publish_volume_info = 3;
+
+  // The path to which the volume will be published. It MUST be an
+  // absolute path in the root filesystem of the process serving this
+  // request. The CO SHALL ensure uniqueness of staging_target_path per
+  // volume.
+  // This is a REQUIRED field.
+  string staging_target_path = 4;
+
+  // The capability of the volume the CO expects the volume to have.
+  // This is a REQUIRED field.
+  VolumeCapability volume_capability = 5;
+
+  // Credentials used by Node plugin to authenticate/authorize node
+  // stage request.
+  // This field contains credential data, for example username and
+  // password. Each key must consist of alphanumeric characters, '-',
+  // '_' or '.'. Each value MUST contain a valid string. An SP MAY
+  // choose to accept binary (non-string) data by using a binary-to-text
+  // encoding scheme, like base64. An SP SHALL advertise the
+  // requirements for credentials in documentation. COs SHALL permit
+  // passing through the required credentials. This information is
+  // sensitive and MUST be treated as such (not logged, etc.) by the CO.
+  // This field is OPTIONAL.
+  map<string, string> node_stage_credentials = 6;
+
+  // Attributes of the volume to publish. This field is OPTIONAL and
+  // MUST match the attributes of the VolumeInfo identified by
+  // `volume_id`.
+  map<string,string> volume_attributes = 7;
+}
+
+message NodeStageVolumeResponse {
+  // The SP specific information that will be passed to the Plugin in
+  // the subsequent `NodePublishVolume` calls
+  // for the given volume.
+  // This information is opaque to the CO. This field is OPTIONAL.
+  map<string, string> publish_volume_info = 1;
```

Comment:
Maybe some clarifying language? I think there might be some confusion now that publish_volume_info is emitted from potentially two sources (ControllerPublishVolumeResponse, and NodeStageVolumeResponse). Is the CO expected to combine the results from both calls when making a subsequent NodePublishVolume call? If this field is not present in this message, but it was in a ControllerPublishVolumeResponse, does that mean that we DO or DO NOT specify a publish_volume_info when calling NodeStageVolumeResponse?

Reply:
Can you explain why this is needed? I'd like to see a concrete use case why this extra information returned from `NodeStageVolume` is necessary.

Reply:
Thought about it some more and discussed with Saad, between the two of us we couldn't actually come up with a concrete use case. I think we'll omit this field, if it turns out there is a use case for it in the future it can be added later.